File: Daily Challenge/April 2025/Lowest Common Ancestor of Deepest Leaves.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
  public TreeNode LcaDeepestLeaves(TreeNode root) {
    if (root == null) return null;
    
    var (depth, lca) = Dfs(root);
    
    return lca;
  }

  private (int, TreeNode) Dfs(TreeNode node) {
    if (node == null) return (0, null);
    
    var (leftDepth, leftLca) = Dfs(node.left);
    var (rightDepth, rightLca) = Dfs(node.right);
    
    if (leftDepth == rightDepth) return (leftDepth + 1, node);
    
    return leftDepth > rightDepth ? (leftDepth + 1, leftLca) : (rightDepth + 1, rightLca);
  }
}

File: Daily Challenge/April 2025/Lowest Common Ancestor of Deepest Leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 class Solution {
public:
  TreeNode* lcaDeepestLeaves(TreeNode* root) {
    if (!root) return nullptr;
    auto [lca, _] = dfs(root, 0);
    return lca;
  }

private:
  pair<TreeNode*, int> dfs(TreeNode* node, int depth) {
    if (!node) return {nullptr, depth};
    auto [leftLCA, leftDepth] = dfs(node->left, depth + 1);
    auto [rightLCA, rightDepth] = dfs(node->right, depth + 1);
    if (leftDepth == rightDepth) return {node, leftDepth};
    return leftDepth > rightDepth ? make_pair(leftLCA, leftDepth) : make_pair(rightLCA, rightDepth);
  }
};